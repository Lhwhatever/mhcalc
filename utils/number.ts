export function coerceToRange(num: number, min = -Infinity, max = +Infinity): number {
    return Math.min(Math.max(num, min), max)
}
