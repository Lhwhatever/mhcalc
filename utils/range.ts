/**
 * Produces an array of strings that represent consecutive integral ranges.
 * Should not mutate the given list.
 * @param list The lower ends of the ranges, as an array of strictly increasing integers.
 * @returns The array of strings that represent integral ranges.
 */
export function formatRange(list: number[]): string[] {
    if (!list.length) return []

    const result = list.slice(1).reduce(
        (cumul, elem) => {
            const { list, last } = cumul
            list.push(last === elem - 1 ? last.toString() : `${last}-${elem - 1}`)
            return { list, last: elem }
        },
        { list: [] as string[], last: list[0] }
    )

    result.list.push(`${result.last}+`)
    return result.list
}
