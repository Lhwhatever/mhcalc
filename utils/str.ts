/**
 * Renders a number as a string. If the parameter given is undefined, returns the empty string.
 *
 * @param num Number to render as a string. Can be undefined.
 * @returns Number as a string.
 */
export function renderNumber(num?: number): string {
    return num === undefined ? '' : num.toString()
}

/**
 * Tries to parse a string as an integer. If the string is empty, returns undefined.
 * If the string is non-numeric, returns NaN.
 * @param str The string to parse.
 * @param radix The base system to use. Defaults to 10.
 * @returns The integer.
 */
export function asInteger(str: string, radix = 10): number | undefined {
    return str ? parseInt(str, radix) : undefined
}

/**
 * Converts a string of alphabetical and whitespace characters in normal English syntax to kebab-case.
 * @param text The text to convert.
 * @returns The text in kebab-case.
 */
export function textToKebabCase(text: string): string {
    return text.toLowerCase().replace(/ /g, '-')
}
