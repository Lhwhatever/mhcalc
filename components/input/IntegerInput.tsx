import { TextField } from '@material-ui/core'
import React from 'react'
import { coerceToRange } from '../../utils/number'
import { asInteger } from '../../utils/str'
import { InputChangeEvent } from './types'

export interface NumericInputProps {
    onChange: (newValue: number | undefined) => void
    value?: number
    min?: number
    max?: number
}

const IntegerInput = (props: NumericInputProps): JSX.Element => {
    const { onChange, min, max, value, ...other } = props

    const handleChange = (event: InputChangeEvent) => {
        if (event.target.value === '') onChange(undefined)
        if (/^-?\d+$/.test(event.target.value)) onChange(asInteger(event.target.value))
    }

    const handleUnfocus = () => {
        if (value === undefined || value < min || value > max) onChange(coerceToRange(value ?? 0, min, max))
    }

    return (
        <TextField type="number" {...other} onChange={handleChange} onBlur={handleUnfocus} inputProps={{ min, max }} />
    )
}

export default IntegerInput
