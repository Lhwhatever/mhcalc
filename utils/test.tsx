import { ThemeProvider } from '@material-ui/core'
import { render as rtlRender, RenderOptions, RenderResult as RtlRenderResult } from '@testing-library/react'
import React from 'react'
import { InputChangeEvent } from '../components/input/types'
import theme from '../theme'

interface ComponentWrapperProps {
    children: React.ReactNode
}

const ComponentWrapper = ({ children }: ComponentWrapperProps) => {
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const render = (ui: React.ReactElement, options?: RenderOptions): RtlRenderResult =>
    rtlRender(ui, { wrapper: ComponentWrapper as React.ComponentType, ...options })

export * from '@testing-library/react'
export { render }

export type RenderResult<P> = Omit<ReturnType<typeof render>, 'rerender'> & {
    rerender: (additionalProps?: Partial<P>) => void
}

type SetupFunc<P> = (additionalProps?: Partial<P>) => RenderResult<P>

export function createSetup<P>(Component: React.ComponentType<P>, componentProps: P): SetupFunc<P> {
    return (additionalProps?: Partial<P>) => {
        const { rerender: oldRerender, ...utils } = render(<Component {...componentProps} {...additionalProps} />)
        return {
            ...utils,
            rerender(additionalProps?: Partial<P>) {
                oldRerender(<Component {...componentProps} {...additionalProps} />)
            }
        }
    }
}

export interface InputChangeEventOptions {
    target: Pick<InputChangeEvent['target'], 'value'>
}

export const createChangeEvent = (value: string): InputChangeEventOptions => ({ target: { value } })
