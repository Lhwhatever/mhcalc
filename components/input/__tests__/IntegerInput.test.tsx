import * as number from '../../../utils/number'
import { createSetup, fireEvent, screen } from '../../../utils/testing/test'
import { createChangeEvent } from '../../../utils/testing/event'
import IntegerInput, { IntegerInputProps } from '../IntegerInput'

describe('IntegerInput test', () => {
    const coerceToRangeSpy = jest.spyOn(number, 'coerceToRange')

    const props: IntegerInputProps = {
        label: 'foo',
        id: 'foo',
        onChange: jest.fn()
    }

    const setup = createSetup(IntegerInput, props)

    const getInput = () => screen.getByLabelText('foo')

    it('should have a label', () => {
        setup()
        const input = getInput()

        expect(input).toBeInTheDocument()
    })

    it('should use type number', () => {
        setup()
        const input = getInput()

        expect(input).toHaveProperty('type', 'number')
    })

    it('should accept integral or empty input', () => {
        setup()
        const input = getInput()

        fireEvent.change(input, createChangeEvent('42'))
        expect(props.onChange).lastCalledWith(42)

        fireEvent.change(input, createChangeEvent('-1'))
        expect(props.onChange).lastCalledWith(-1)
    })

    it('should reject non-integral input', () => {
        setup()
        const input = getInput()

        fireEvent.change(input, createChangeEvent('1a'))
        expect(props.onChange).not.toBeCalled()

        fireEvent.change(input, createChangeEvent('1.5'))
        expect(props.onChange).not.toBeCalled()
    })

    it('should pass the min and max properties to input', () => {
        const { rerender } = setup({ min: 1 })
        const input = getInput()

        expect(input).toHaveAttribute('min', '1')
        expect(input).not.toHaveAttribute('max')

        rerender({ max: 5 })
        expect(input).not.toHaveAttribute('min')
        expect(input).toHaveAttribute('max', '5')

        rerender({ min: 1, max: 5 })
        expect(input).toHaveAttribute('min', '1')
        expect(input).toHaveAttribute('max', '5')
    })

    it('should, on blur, try to coerce values into the given range', () => {
        setup({ value: -1, min: 1, max: 5 })
        const input = getInput()

        fireEvent.blur(input)
        expect(coerceToRangeSpy).lastCalledWith(-1, 1, 5)
        expect(props.onChange).toBeCalledWith(coerceToRangeSpy.getMockImplementation()(-1, 1, 5))
    })

    it('should, on blur, try to coerce undefined as zero', () => {
        setup({ value: undefined, min: -5, max: 5 })
        const input = getInput()

        fireEvent.blur(input)
        expect(coerceToRangeSpy).toBeCalledWith(0, -5, 5)
        expect(props.onChange).toBeCalledWith(coerceToRangeSpy.getMockImplementation()(0, -5, 5))
    })

    it('should be uncontrolled if no onChange is given', () => {
        setup({ onChange: undefined })
        const input = getInput()
        fireEvent.change(input, createChangeEvent('5'))
        expect(input).toHaveValue(null)
    })

    it('should match snapshot', () => {
        setup()
        expect(screen).toMatchSnapshot()
    })
})
