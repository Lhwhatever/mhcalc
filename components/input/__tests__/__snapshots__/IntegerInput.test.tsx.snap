// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericInput test should match snapshot 1`] = `
<WithStyles(ForwardRef(TextField))
  inputProps={
    Object {
      "max": undefined,
      "min": undefined,
    }
  }
  onBlur={[Function]}
  onChange={[Function]}
  type="number"
/>
`;
